### PR TITLE
add benmark plotting for constrained optimizer

### PR DESCRIPTION
**Pull request by yetongumich**
Add constrained optimization, with example plots:
Figure_1
Figure_2

### CODE REVIEW

File: gtdynamics/optimizer/AugmentedLagrangianOptimizer.cpp [hunk 1]
```diff
@@ -47,7 +47,8 @@ void update_parameters(const EqualityConstraints& constraints,
 gtsam::Values AugmentedLagrangianOptimizer::optimize(
     const gtsam::NonlinearFactorGraph& graph,
     const EqualityConstraints& constraints,
-    const gtsam::Values& initial_values) const {
+    const gtsam::Values& initial_values,
+    boost::optional<ConstrainedOptResult&> opt_result) const {
```
Comment (dellaert) [inline gtdynamics/optimizer/AugmentedLagrangianOptimizer.cpp:51]: Google style says: in/out parameters should be pointers. Pls use `ConstrainedOptResult*a` and pass in &results…

**Review by dellaert — APPROVED** [review]
Coool!! But please use pointer for in/out.

File: gtdynamics/optimizer/AugmentedLagrangianOptimizer.cpp (hunk 1 above)
Comment (yetongumich) [reply gtdynamics/optimizer/AugmentedLagrangianOptimizer.cpp]: Done

File: gtdynamics/optimizer/AugmentedLagrangianOptimizer.cpp [hunk 2]
```diff
@@ -47,7 +47,8 @@ void update_parameters(const EqualityConstraints& constraints,
 gtsam::Values AugmentedLagrangianOptimizer::optimize(
     const gtsam::NonlinearFactorGraph& graph,
     const EqualityConstraints& constraints,
-    const gtsam::Values& initial_values) const {
+    const gtsam::Values& initial_values,
+    boost::optional<ConstrainedOptResult*> opt_result) const {
```
Comment (dellaert) [inline gtdynamics/optimizer/AugmentedLagrangianOptimizer.cpp:51]: No, sorry, this should just be `ConstrainedOptResult*`. You can check if it is `!= nullptr` and fill it if not. Finally, if you keep this parameter, just call it `result`. It's optional and i/o by virtue of being a non-const pointer.

File: gtdynamics/optimizer/ConstrainedOptimizer.h [hunk 3]
```diff
@@ -30,6 +30,13 @@ struct ConstrainedOptimizationParameters {
   ConstrainedOptimizationParameters() {}
 };
 
+/// Intermediate results for constrained optimization process.
+struct ConstrainedOptResult {
+  std::vector<gtsam::Values> values_list;  // values after each inner loop
```
Comment (dellaert) [inline gtdynamics/optimizer/ConstrainedOptimizer.h:35]: never put type in a variable name, and definitely not the wrong type :-) Just say `intermediate_values`

File: gtdynamics/optimizer/ConstrainedOptimizer.h [hunk 4]
```diff
@@ -30,6 +30,13 @@ struct ConstrainedOptimizationParameters {
...
+  std::vector<int> num_iters;   // number of LM iterations for each inner loop
+  std::vector<double> mu_list;  // penalty parameter for each inner loop
```
Comment (dellaert) [inline gtdynamics/optimizer/ConstrainedOptimizer.h:37]: `mu_values`

File: scripts/nithya_yetong00_constrainedopt_benchmark.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
+/* ----------------------------------------------------------------------------
+ * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * See LICENSE for the license information
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file  nithya_yetong00_constrainedopt_benchmark.cpp
```
Comment (dellaert) [inline scripts/nithya_yetong00_constrainedopt_benchmark.cpp:9]: Just nithya :-) And, `constrained_opt_benchmark`.

**Review by dellaert — APPROVED** [review]
Some more comments but looks good otherwise.

File: gtdynamics/optimizer/ConstrainedOptimizer.h (hunk 3 above)
Comment (yetongumich) [reply gtdynamics/optimizer/ConstrainedOptimizer.h:35]: Done

File: gtdynamics/optimizer/ConstrainedOptimizer.h (hunk 4 above)
Comment (yetongumich) [reply gtdynamics/optimizer/ConstrainedOptimizer.h:37]: Done

File: scripts/nithya_yetong00_constrainedopt_benchmark.cpp (hunk 5 above)
Comment (yetongumich) [reply scripts/nithya_yetong00_constrainedopt_benchmark.cpp:9]: Done

File: gtdynamics/optimizer/AugmentedLagrangianOptimizer.cpp (hunk 2 above)
Comment (yetongumich) [reply gtdynamics/optimizer/AugmentedLagrangianOptimizer.cpp:51]: Done

**Pull request merged**